control/state.py

```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
-    def _restore_listeners(self, omap_dict, callback):
-        """Restores listeners from the OMAP."""
-
-        for (key, val) in omap_dict.items():
-            if key.startswith(self.LISTENER_PREFIX):
-                req = json_format.Parse(val, pb2.create_listener_req())
-                callback(req)
-
-    def _read_key(self, key) -> Optional[str]:
-        """Reads a key from the OMAP and returns its value."""
-
-        with rados.ReadOpCtx() as read_op:
-            iter, _ = self.ioctx.get_omap_vals_by_keys(read_op, (key,))
-            self.ioctx.operate_read_op(read_op, self.omap_name)
-            value_list = list(dict(iter).values())
-            if len(value_list) == 1:
-                val = str(value_list[0], "utf-8")
-                self.logger.debug(f"Read key: {key} -> {val}")
-                return val
-        return None
-
-    def _read_all(self) -> Dict[str, str]:
-        """Reads OMAP and returns dict of all keys and values."""
-
-        with rados.ReadOpCtx() as read_op:
-            iter, _ = self.ioctx.get_omap_vals(read_op, "", "", -1)
-            self.ioctx.operate_read_op(read_op, self.omap_name)
-            omap_dict = dict(iter)
-        return omap_dict
+        """Removes a listener from the state data store."""
+        self.remote.remove_listener(subsystem_nqn, gateway, trtype, traddr,
+                                    trsvcid)
+        self.local.remove_listener(subsystem_nqn, gateway, trtype, traddr,
+                                   trsvcid)
 
     def delete_state(self):
-        """Deletes OMAP object."""
-
-        try:
-            self.ioctx.remove_object(self.omap_name)
-            self.logger.info(f"Object {self.omap_name} deleted.")
-        except rados.ObjectNotFound:
-            self.logger.info(f"Object {self.omap_name} not found.")
-
-    def restore(self, callbacks):
-        """Restores gateway state to OMAP specifications."""
-
-        omap_version = self._read_key(self.OMAP_VERSION_KEY)
-        if omap_version == "1":
-            self.logger.info("This omap was just created. Nothing to restore")
+        """Deletes state data stores."""
+        self.remote.delete_state()
+        self.local.delete_state()
+
+    def start_update(self, gateway_rpc_caller):
+        self.gateway_rpc_caller = gateway_rpc_caller
+
+        # Update to restore existing state on restart
+        self.update()
+
+        if self.config.getboolean("gateway", "state_update_notify"):
+            # Register a watch on remote state
+            self.remote.register_watch(self.update)
+
+        # Start polling for state updates
+        self.update_interval = self.config.getint("gateway",
+                                                  "state_update_interval_sec")
+        if self.update_interval < 1:
+            self.logger.info("Invalid state_update_interval_sec. Setting to 1.")
+            self.update_interval = 1
+        self.update_timer = threading.Thread(target=self._update_caller)
```

Comment:
Define variable in `__init__` and set to `None`. Check if already set so in case function gets called again we don't get a 2nd thread.